crates/subspace-networking/src/connected_peers.rs

```diff
@@ -137,61 +140,60 @@ impl Behaviour {
             dialing_delay,
             peer_cache: Vec::new(),
             waker: None,
+            peer_connection_counters: HashMap::new(),
         }
     }
 
     /// Create a connection handler for the protocol.
-    #[inline]
     fn new_connection_handler(&mut self, peer_id: &PeerId) -> Handler {
         let default_until = Instant::now().add(self.config.decision_timeout);
         let default_keep_alive_until = KeepAlive::Until(default_until);
         let keep_alive = if let Some(decision) = self.known_peers.get_mut(peer_id) {
             match decision {
-                PeerDecision::PendingConnection { .. } => {
+                PeerConnectionState::Connecting { .. } => {
                     // Connection attempt was successful.
-                    *decision = PeerDecision::PendingDecision {
+                    *decision = PeerConnectionState::Deciding {
                         until: default_until,
                     };
 
                     default_keep_alive_until
                 }
-                PeerDecision::PendingDecision { until } => KeepAlive::Until(*until),
-                PeerDecision::PermanentConnection => KeepAlive::Yes,
-                PeerDecision::NotInterested => KeepAlive::No,
+                PeerConnectionState::Deciding { until } => KeepAlive::Until(*until),
+                PeerConnectionState::Permanent => KeepAlive::Yes,
+                PeerConnectionState::NotInterested => KeepAlive::No,
             }
         } else {
-            default_keep_alive_until
-        };
-
-        // Connection from other protocols.
-        if !self.known_peers.contains_key(peer_id) {
+            // Connection from other protocols.
             self.known_peers.insert(
                 *peer_id,
-                PeerDecision::PendingDecision {
+                PeerConnectionState::Deciding {
                     until: default_until,
                 },
             );
-        }
+
+            default_keep_alive_until
+        };
 
         self.wake();
         Handler::new(self.config.protocol_name, keep_alive)
     }
 
-    /// Specifies the whether we should keep connections to the peer alive.
-    /// Note: in rare cases, it could override decision timeout and keep connection alive.
+    /// Specifies the whether we should keep connections to the peer alive. The decision could
+    /// depend on another protocol (e.g.: PeerInfo protocol event handling).
+    /// In case when we had decision timeout it sets up proper keep connection alive anyway.
     pub fn update_keep_alive_status(&mut self, peer_id: PeerId, keep_alive: bool) {
```

Comment:
Why do we have `keep_alive: bool` as an argument, in which scenario do you expect that we need to change decision from interested to not interested?

Reply:
Discussed offline, we could reject a new connection as 'NotInterested` and thus reduce the interval of live connection in opposite of waiting of connection timeout.

Reply:
My concern was slightly different. Current API allows you to change state from `ConnectionState::Permanent` to `ConnectionState::NotInterested`, which doesn't seem like a thing we want to expose.